src/index_intf.ml

```diff
@@ -0,0 +1,271 @@
+  (** [with_lock t f] first obtains [t], then computes [f ()], and finally
+      unlocks [t]. *)
+end
+
+module type THREAD = sig
+  (** Cooperative threads. *)
+
+  type 'a t
+  (** The type of thread handles. *)
+
+  val async : (unit -> 'a) -> 'a t
+  (** [async f] creates a new thread of control which executes [f ()] and
+      returns the corresponding thread handle. The thread terminates whenever
+      [f ()] returns a value or raises an exception. *)
+
+  val await : 'a t -> ('a, [ `Async_exn of exn ]) result
+  (** [await t] blocks on the termination of [t]. *)
+
+  val return : 'a -> 'a t
+  (** [return ()] is a pre-terminated thread handle. *)
+
+  val yield : unit -> unit
+  (** Re-schedule the calling thread without suspending it. *)
+end
+
+module type S = sig
+  type t
+  (** The type for indexes. *)
+
+  type key
+  (** The type for keys. *)
+
+  type value
+  (** The type for values. *)
+
+  val v : ?fresh:bool -> ?readonly:bool -> log_size:int -> string -> t
+  (** The constructor for indexes.
+
+      @param fresh whether an existing index should be overwritten.
+      @param read_only whether read-only mode is enabled for this index.
+      @param log_size the maximum number of bindings in the `log` IO. *)
+
+  val clear : t -> unit
+  (** [clear t] clears [t] so that there are no more bindings in it. *)
+
+  val find : t -> key -> value
+  (** [find t k] is the binding of [k] in [t]. *)
+
+  val mem : t -> key -> bool
+  (** [mem t k] is [true] iff [k] is bound in [t]. *)
+
+  exception Invalid_key_size of key
+
+  exception Invalid_value_size of value
+  (** The exceptions raised when trying to add a key or a value of different
+      size than encoded_size *)
+
+  val replace : t -> key -> value -> unit
+  (** [replace t k v] binds [k] to [v] in [t], replacing any existing binding of
+      [k]. *)
+
+  val filter : t -> (key * value -> bool) -> unit
+  (** [filter t p] removes all the bindings (k, v) that do not satisfy [p]. This
+      operation is costly and blocking. *)
+
+  val iter : (key -> value -> unit) -> t -> unit
+  (** Iterates over the index bindings. Limitations:
+
+      - Order is not specified.
+      - In case of recent replacements of existing values (since the last
+        merge), this will hit both the new and old bindings.
+      - May not observe recent concurrent updates to the index by other
+        processes. *)
+
+  val flush : ?with_fsync:bool -> t -> unit
+  (** Flushes all internal buffers of the [IO] instances. If [with_fsync] is
+      [true], this also flushes the OS caches for each [IO] instance. *)
+
+  val close : t -> unit
+  (** Closes all resources used by [t]. *)
+end
+
+module type Index = sig
+  (** The input of [Make] for keys. *)
+  module type Key = sig
+    (* N.B. We use [sig ... end] redirections to avoid linking to the [_intf]
+       file in the generated docs. Once Odoc 2 is released, this can be
+       removed. *)
+
+    include Key
+    (** @inline *)
+  end
```

Comment:
Here's the gymnastics in question. It's sufficient to do:

```ocaml
module type Key = Key
```

but then Odoc will add an extra link indirection through the `_intf` file: not a great UX.